modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/sys/view/SqlSystemViewColumns.java

```diff
@@ -0,0 +1,142 @@
+
+/**
+ * View that contains information about all the sql tables' columns in the cluster.
+ */
+public class SqlSystemViewColumns extends SqlAbstractLocalSystemView {
+    /** Name of the column that contains names of sql tables. */
+    private static final String TABLE_NAME = "TABLE_NAME";
+
+    /** Name of the column that contains names of sql table's columns. */
+    private static final String COLUMN_NAME = "COLUMN_NAME";
+
+    /** Schema manager. */
+    private final SchemaManager schemaMgr;
+
+    /**
+     * Creates view with columns.
+     *
+     * @param ctx Kernal context.
+     * @param schemaMgr Schema manager.
+     */
+    public SqlSystemViewColumns(GridKernalContext ctx, SchemaManager schemaMgr) {
+        super("COLUMNS", "Ignite columns", ctx, new String[] {TABLE_NAME, COLUMN_NAME},
+            newColumn("SCHEMA_NAME"),
+            newColumn(TABLE_NAME),
+            newColumn(COLUMN_NAME),
+            newColumn("COLUMN_ID", Value.INT),
+            newColumn("DATA_TYPE"),
+            newColumn("SQL_TYPE", Value.INT),
+            newColumn("IS_NULLABLE", Value.BOOLEAN),
+            newColumn("DISPLAY_SIZE", Value.INT),
+            newColumn("PRECISION", Value.INT),
+            newColumn("SCALE", Value.INT),
+            newColumn("AFFINITY_KEY", Value.BOOLEAN)
+        );
+
+        this.schemaMgr = schemaMgr;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Iterator<Row> getRows(Session ses, SearchRow first, SearchRow last) {
+        SqlSystemViewColumnCondition tableNameCond = conditionForColumn(TABLE_NAME, first, last);
+        SqlSystemViewColumnCondition columnNameCond = conditionForColumn(COLUMN_NAME, first, last);
+
+        Predicate<GridH2Table> tblFilter;
+        Predicate<Column> colFilter;
+
+        if (tableNameCond.isEquality()) {
+            String tblName = tableNameCond.valueForEquality().getString();
+
+            tblFilter = tbl -> F.eq(tblName, tbl.getName());
+        }
+        else
+            tblFilter = tab -> true;
+
+        if (columnNameCond.isEquality()) {
+            String colName = columnNameCond.valueForEquality().getString();
+
+            colFilter = col -> F.eq(colName, col.getName());
+        }
+        else
+            colFilter = tab -> true;
+
+        List<Row> res = new ArrayList<>();
+
+        for (GridH2Table tbl : schemaMgr.dataTables()) {
+            if (!tblFilter.test(tbl))
+                continue;
+
+            IndexColumn affCol = tbl.getExplicitAffinityKeyColumn();
+
+            for (int i = QueryUtils.DEFAULT_COLUMNS_COUNT; i < tbl.getColumns().length; ++i) {
+                Column col = tbl.getColumns()[i];
+
+                if (!colFilter.test(col))
+                    continue;
+
+                Object[] data = new Object[] {
+                    col.getTable().getSchema().getName(),
+                    col.getTable().getName(),
+                    col.getName(),
+                    col.getColumnId(),
```

Comment:
I'm not sure that we should use here internal column id, may be will be better to use col.getColumnId() - QueryUtils.DEFAULT_COLUMNS_COUNT as value.